private/pkg/git/url.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package git
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/bufbuild/buf/private/pkg/command"
+)
+
+const (
+	githubGitlabRemoteURLFormat = "https://%s%s/commit/%s"
+	bitBucketRemoteURLFormat    = "https://%s%s/commits/%s"
+	bitBucketHostname           = "bitbucket"
+	githubHostname              = "github"
+	gitlabHostname              = "gitlab"
+)
+
+// GetSourceControlURL gets the parses a user-facing source control URL based on the URL
+// from git config based on the given remote, git commit sha, and directory.
+//
+// It checks the hostname of repository remote URL -- if it contains a known hostname,
+// github, gitlab, and bitbucket, we construct a URL. Otherwise, we return an empty source
+// control URL.
+func GetSourceControlURL(
```

Comment:
So I waffled a little on this one -- but since these functions are not encapsulated by an interface, I ended up not adding it to `git.go` because I didn't want to conflate the two styles... I can move them to `git.go` if they are confusing.